### DISCUSSION TITLE
Is Redis really necessary?

### DISCUSSION THREAD

**Discussion by dudicoco** [argoproj/argo-cd#6618] [Q&A] [answered]
Hi,

There is very little information within the docs regarding the necessity of using redis for caching.

According to https://argo-cd.readthedocs.io/en/stable/operator-manual/architecture/#repository-server:
```
The repository server is an internal service which maintains a local cache of the Git repository holding the application manifests.
```

If the repo server already maintains a local cache, why do we need redis in addition? 
Can we just convert the repo server to a statefulset to maintain the cache during restarts and remove the need for redis?

Thanks

**Comment by exe-r** [comment]
Tried to deploy argocd using helm (latest charts) with redis disabled but I was unable to get it working past SSO login is giving error: failed to verify app state eDJFHSJS: dial tcp: lookup argocd-redis on x.x.x.x:53: server misbehaving. So I ended up with the same question, is redis a requirement or not? if it is it should be clearly stated on the docs IMO.

**Comment by jannfis** [comment, marked as answer]
Running Argo CD without Redis is not supported, as it is an integral part of the architecture. You can opt to use your own Redis instance however.

As for a little background info, we do cache a lot of things, not only local clones of Git repositories - for example, Kubernetes resources, connection states, etc. We treat the cache is volatile, that means, we don't persist anything in Redis and just rebuild the cache after it has been flushed.

**Reply by exe-r** [reply]
Thanks for the clarification on this.

**Reply by emirot** [reply]
@jannfis does it mean that if let's say redis pod crashed, and then restart, would argoCD still work but will be slower ? (during the time of recreating cache entries)
Also would missing key entry will cause argocd not to work? I m curious since I've seen [that thread ](https://github.com/argoproj/argo-cd/issues/9622#issuecomment-1189103756)?

**Reply by jannfis** [reply]
@emirot If a single instance Redis cache crashes and needs to be restarted, most likely you would see some intermittent errors which would require a (possibly manual) sync or refresh of the applications.

For this reason, Redis can be installed in HA mode (with one primary and multiple secondaries). This is part of the HA installation manifests.

**Reply by emirot** [reply]
Maybe the doc https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/high_availability.md#high-availability will need to be updated from the doc it looks like nothing will be impacted.

**Reply by litan-shamir** [reply]
Hello @jannfis,

We attempted to reduce costs by downsizing the redis-ha replicas to one, which would eliminate data replication across availability zones. Unfortunately, we encountered some unexpected behavior, and the resources created by resources are not displaying correctly.

<img width="538" alt="image" src="https://user-images.githubusercontent.com/123365766/232322386-4cb8738b-67d0-4686-afa6-9388ca4e65df.png">

<img width="880" alt="image" src="https://user-images.githubusercontent.com/123365766/232322466-05306570-f3cd-4e17-8040-d24bdc648799.png">


Can you please confirm if this is a viable approach to reducing replication costs? Also, would you happen to have any suggestions on avoiding replication costs without downsizing the redis-ha replicas? Perhaps by deploying the redis-ha on specific AZ nodes?

Thank you for being so supportive.

**Reply by javidaslan** [reply]
Hi @litan-shamir ,

Just came across your question while investigation how important the Redis for ArgoCD is. We were suffering from high cross-AZ costs due to the replication too. For us solution was to restrict Redis deployment to the single AZ while keeping `podAntiAffinity` in place, so that pods would still get scheduled onto different nodes (to provide some high availability). 

I know that with this approach AZ level high-availability is compromised, but in our situation, we could accept this trade-off.

I hope it helps :)

**Comment by manish222agr** [comment]
What setup or configuration required for azure redis for argocd to run. If I do not make any configurations, is argocd expected to create redis  by itself. 
If so, where. I am deploying argocd to azure kuberenetes cluster.
